src/lib/Parser.hs

```diff
@@ -285,6 +335,41 @@ decl = do
   rhs <- sym "=" >> blockOrExpr
   return $ lhs rhs
 
+interfaceInstance :: Parser UDecl
+interfaceInstance = do
+  keyWord InstanceKW
+  (p, pos) <- withPos letPat
+  ann <- annot uType
+  keyWord WhereKW
+  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
+  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+  return $ ULet InstanceLet (p, ann') rhs'
+  where
+    -- Here, we are traversing the type annotation to retrieve the name of
+    -- the interface and generate its corresponding constructor.
+    mkRhs (WithSrc _ (UPi _ _ lrhs))      rhs = mkRhs lrhs rhs
+    mkRhs (WithSrc _ (UApp _ (WithSrc _ (UVar v)) _)) rhs =
+      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
+    -- We let this case through to fail during type checking.
+    mkRhs _ rhs = rhs
```

Comment:
This seems like a strange fallback case to me? In particular, if we can't figure out what the RHS of the instance should be, we just use a record directly?

What kinds of errors does this produce if you do it wrong? I'd rather have a clear error during parsing than a confusing error during type inference.

Reply:
Thanks for your comments @danieldjohnson! I'll look at the rest later (about to head offline), but this in particular we discussed a little bit here (I don't know if you saw it): https://github.com/google-research/dex-lang/pull/249#discussion_r498738453

Reply:
I didn't see that! But I think the "good" error message in that case is a kind error, because the declared annotation wasn't a type. If you give it a valid type that just isn't a interface type, you get a different error:
```
instance nonsensical : Int32 where
  foo = 3
> Type error:
> Expected: Int32
>   Actual: {foo: Int32}
>
> instance nonsensical : Int32 where
>                                   ^
```
Which... isn't perfect but I guess isn't that bad? Maybe a non-intended side effect is that this works:
```
instance myrecord : {a:Int & b:Float} where
  a = 3
  b = 42.0

:p myrecord
> {a = 3, b = 42.0}
```

Reply:
On the subject of error messages when people do things incorrectly, here's another case that could probably be clearer (the error root cause here is that the annotation uses `->` instead of `?=>`):
```
instance tabAddIncorrect : Add a -> Add (n=>a) where
  add = \xs ys. for i. xs.i + ys.i
  sub = \xs ys. for i. xs.i - ys.i
  zero = for _. zero
> Type error:
> Expected: ((Add a) -> Add (n => a))
>   Actual: (Add (b => a))
> (Solving for: [a:Type, b:Type])
>
> instance tabAddIncorrect : Add a -> Add (n=>a) where
>                                                     ^
```